File: examples/74-util-api/main.go
// examples/74-util-api demonstrates unit conversion helpers.
//
// Shows the styling.Length type and the Inches, Centimeters, Points, and Emu
// constructor functions, plus the reverse conversion methods (.Inches(), .Cm(),
// .Pt(), .Emu()) and FontSize().
//
// Run with: go run ./examples/74-util-api/main.go
package main

import (
	"fmt"
	"os"
	"path/filepath"

	log "github.com/djinn-soul/gopptx/pkg/stdlog"

	"github.com/djinn-soul/gopptx/pkg/pptx"
	"github.com/djinn-soul/gopptx/pkg/pptx/styling"
)

const (
	outputDir  = "examples/output"
	outputFile = "74_util_api.pptx"
)

func main() {
	if err := run(); err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}
}

func run() error {
	if err := os.MkdirAll(outputDir, 0o750); err != nil {
		return fmt.Errorf("create output directory: %w", err)
	}

	// --- Demonstrate unit conversions ---

	// Inches <-> EMU
	oneInch := styling.Inches(1)
	log.Printf("1 inch = %d EMU\n", oneInch.Emu())
	log.Printf("914400 EMU = %.4f inches\n", styling.Emu(914400).Inches())

	// Centimeters <-> EMU
	oneCm := styling.Centimeters(1)
	log.Printf("1 cm = %d EMU\n", oneCm.Emu())
	log.Printf("360000 EMU = %.4f cm\n", styling.Emu(360000).Cm())

	// Points <-> EMU
	onePt := styling.Points(1)
	log.Printf("1 pt = %d EMU\n", onePt.Emu())
	log.Printf("12700 EMU = %.4f pt\n", styling.Emu(12700).Pt())

	// Direct EMU
	directEmu := styling.Emu(914400)
	log.Printf("Emu(914400) inches = %.4f\n", directEmu.Inches())

	// FontSize (pt -> hundredths-of-a-point for OOXML)
	log.Printf("FontSize(18pt) = %d (hundredths of a point)\n", styling.FontSize(18))
	log.Printf("FontSize(44pt) = %d\n", styling.FontSize(44))

	// MaxEMU boundary
	log.Printf("MaxEMU = %d\n", styling.MaxEMU.Emu())

	// --- Build a slide using every unit type ---
	rectInches := pptx.NewRectangle(
		styling.Inches(0.5).Inches(),
		styling.Inches(1.5).Inches(),
		styling.Inches(2.5).Inches(),
		styling.Inches(1).Inches(),
	).WithFill(pptx.NewShapeFill("4472C4")).WithText("Inches")

	rectCm := pptx.NewShape(
		pptx.ShapeTypeRectangle,
		styling.Centimeters(8),
		styling.Centimeters(4),
		styling.Centimeters(5),
		styling.Centimeters(2),
	).WithFill(pptx.NewShapeFill("C0504D")).WithText("Centimeters")

	rectPt := pptx.NewShape(
		pptx.ShapeTypeRectangle,
		styling.Points(50),
		styling.Points(400),
		styling.Points(200),
		styling.Points(80),
	).WithFill(pptx.NewShapeFill("9BBB59")).WithText("Points")

	rectEmu := pptx.NewShape(
		pptx.ShapeTypeRectangle,
		styling.Emu(5000000),
		styling.Emu(3000000),
		styling.Emu(2500000),
		styling.Emu(800000),
	).WithFill(pptx.NewShapeFill("F79646")).WithText("EMU")

	infoSlide := pptx.NewSlide("Unit Conversion Results").
		AddBullet(fmt.Sprintf("1 inch = %d EMU", oneInch.Emu())).
		AddBullet(fmt.Sprintf("1 cm   = %d EMU", oneCm.Emu())).
		AddBullet(fmt.Sprintf("1 pt   = %d EMU", onePt.Emu())).
		AddBullet(fmt.Sprintf("FontSize(18pt) = %d", styling.FontSize(18))).
		AddBullet(fmt.Sprintf("MaxEMU = %d", styling.MaxEMU.Emu()))

	shapesSlide := pptx.NewSlide("Shapes Placed with Different Units").
		AddShape(rectInches).
		AddShape(rectCm).
		AddShape(rectPt).
		AddShape(rectEmu)

	slides := []pptx.SlideContent{infoSlide, shapesSlide}

	outputPath := filepath.Join(outputDir, outputFile)
	data, err := pptx.CreateWithSlides("Util API Demo", slides)
	if err != nil {
		return fmt.Errorf("create presentation: %w", err)
	}
	if err := os.WriteFile(outputPath, data, 0o600); err != nil {
		return fmt.Errorf("write file: %w", err)
	}

	log.Printf("Generated %s\n", outputPath)
	return nil
}

File: examples/74-util-api/main.py
"""Demonstrate unit conversion helpers and the Inches() utility.

This example demonstrates:
- Inches() for positioning and sizing shapes
- Placement of shapes using Inches() for different coordinates
- get_slide_size() to read the current slide dimensions
- set_slide_size() to change the presentation canvas size
- The SIZE_16X9_WIDTH / SIZE_16X9_HEIGHT constants
"""

from __future__ import annotations

from pathlib import Path

from gopptx import Presentation
from gopptx.constants import (
    SIZE_16X9_HEIGHT,
    SIZE_16X9_WIDTH,
    ShapeType,
)
from gopptx.presentation.slides import SlideLayoutType
from gopptx.schemas import Inches


def _add_unit_info_slide(prs: Presentation) -> None:
    """Slide listing unit conversion facts."""
    prs.add_bullet_slide(
        "Unit Conversion Reference",
        [
            "Inches(1)       = 914400 EMU",
            "Inches(2.54 cm) = 914400 EMU (1 inch)",
            "1 point         = 12700 EMU",
            "Inches() returns an integer EMU value for OOXML",
            f"SIZE_16X9_WIDTH  = {SIZE_16X9_WIDTH} EMU",
            f"SIZE_16X9_HEIGHT = {SIZE_16X9_HEIGHT} EMU",
        ],
    )


def _add_shapes_with_units_slide(prs: Presentation) -> None:
    """Shapes placed with Inches() for each coordinate system."""
    idx = prs.add_slide(
        "Shapes Placed with Inches()", layout=SlideLayoutType.TITLE_ONLY
    )
    placements = [
        (
            "Inches - top-left",
            "4472C4",
            Inches(0.5),
            Inches(1.5),
            Inches(2.5),
            Inches(1.0),
        ),
        (
            "Inches - centre",
            "C0504D",
            Inches(3.5),
            Inches(1.5),
            Inches(2.5),
            Inches(1.0),
        ),
        (
            "Inches - right",
            "9BBB59",
            Inches(6.5),
            Inches(1.5),
            Inches(2.5),
            Inches(1.0),
        ),
        (
            "Inches - row 2",
            "F79646",
            Inches(0.5),
            Inches(3.0),
            Inches(2.5),
            Inches(1.0),
        ),
    ]
    for label, color, x, y, w, h in placements:
        prs.add_shape(
            idx,
            ShapeType.RECTANGLE,
            (x, y, w, h),
            text=label,
            properties={"fill_color": color},
        )


def _add_slide_size_slide(prs: Presentation) -> None:
    """Demonstrate reading the slide size."""
    size = prs.get_slide_size()
    prs.add_bullet_slide(
        "Slide Size API",
        [
            f"prs.get_slide_size() = {size}",
            "prs.set_slide_size(width_emu, height_emu) changes the canvas",
            f"SIZE_16X9_WIDTH  = {SIZE_16X9_WIDTH}",
            f"SIZE_16X9_HEIGHT = {SIZE_16X9_HEIGHT}",
            "from gopptx.constants import SIZE_16X9_WIDTH, SIZE_16X9_HEIGHT",
        ],
    )


def main() -> None:
    """Create presentation demonstrating Inches() and unit conversion helpers."""
    output_dir = Path("examples/output")
    output_dir.mkdir(exist_ok=True)

    output_path = output_dir / "74_util_api.pptx"

    with Presentation.new("Util API Demo") as prs:
        _add_unit_info_slide(prs)
        _add_shapes_with_units_slide(prs)
        _add_slide_size_slide(prs)

        prs.save(str(output_path))
        print(f"Saved: {output_path}")

    print("\n=== SUMMARY ===")
    print("Demonstrated: Inches(), SIZE_16X9_WIDTH/HEIGHT, get_slide_size(),")
    print("  shape placement with consistent Inches() coordinates")


if __name__ == "__main__":
    main()
